cou/apps/base.py

```diff
@@ -538,59 +527,62 @@ def _get_upgrade_current_release_packages_step(
 
         return step
 
-    def _get_charm_hub_migration(self) -> PreUpgradeStep:
+    def _get_refresh_charm_step(self, target: OpenStackRelease) -> PreUpgradeStep:
+        """Get step for refreshing the charm.
+
+        :param target: OpenStack release as target to upgrade
+        :type target: OpenStackRelease
+        :raises ApplicationError: When application has unexpected channel.
+        :return: Step for refreshing the charm
+        :rtype: PreUpgradeStep
+        """
+        if self.is_from_charm_store:
+            return self._get_charmhub_migration_step()
+        if self.channel == LATEST_STABLE:
+            return self._get_change_to_openstack_channels_step()
+        if self._need_to_refresh(target):
+            return self._get_refresh_current_channel_step()
+        return PreUpgradeStep()
```

Comment:
nit: I think it would be useful to have an info-level log message informing the user that the charm does't need to be refreshed.